### PR TITLE
Add MySQL 8.4 unit tests

### PR DESCRIPTION
**Pull request by dbussink**
## Description

In order to support MySQL 8.4 in the future, we need to run CI unit tests against it. 

Part of https://github.com/vitessio/vitess/issues/16466

## Checklist

-   [x] "Backport to:" labels have been added if this change should be back-ported to release branches
-   [x] If this change is to be back-ported to previous releases, a justification is included in the PR description
-   [x] Tests were added or are not required
-   [x] Did the new or modified tests pass consistently locally and on CI?
-   [x] Documentation was added or is not required

### CODE REVIEW

File: go/mysql/capabilities/capability.go [hunk 1]
```diff
@@ -48,6 +48,8 @@ const (
 	PerformanceSchemaDataLocksTableCapability                           // supported in MySQL 8.0.1 and above: https://dev.mysql.com/doc/relnotes/mysql/8.0/en/news-8-0-1.html
 	InstantDDLXtrabackupCapability                                      // Supported in 8.0.32 and above, solving a MySQL-vs-Xtrabackup bug starting 8.0.29
 	ReplicaTerminologyCapability                                        // Supported in 8.0.26 and above, using SHOW REPLICA STATUS and all variations.
+	RestrictFKOnNonStandardKey                                          // Supported in 8.4.0 and above, restricts usage of non-standard indexes for foreign keys.
+	BinaryLogStatus                                                     // Supported in 8.4.0 and above, uses SHOW BINARY LOG STATUS
```
Comment (GrahamCampbell) [inline go/mysql/capabilities/capability.go:52]: I believe this is 8.2.0 and above, not 8.4.0 and above?

**Review by GrahamCampbell — CHANGES_REQUESTED** [review]

File: go/mysql/capabilities/capability.go (hunk 1 above)
Comment (GrahamCampbell) [reply go/mysql/capabilities/capability.go:52]: See the SQL Syntax Notes section of https://dev.mysql.com/doc/relnotes/mysql/8.2/en/news-8-2-0.html.

Comment (dbussink) [reply go/mysql/capabilities/capability.go:52]: @GrahamCampbell WIP, haven't audited the final additions yet. For Vitess we only really care about MySQL LTS releases anymore. I'll fix once it all works.

Comment (dbussink) [reply go/mysql/capabilities/capability.go]: Fixed up!

**Comment by dbussink** [comment]
Didn't mark this specific one as needing release notes etc, since we only want to add those once we've completed the 8.4 stuff.

**Review by frouioui — COMMENTED** [review]
We should make sure the MySQL version in CI is the same as the one we ship in our docker images as the default MySQL version.

**Comment by dbussink** [comment]
> We should make sure the MySQL version in CI is the same as the one we ship in our docker images as the default MySQL version.

I don't understand this comment? This only adds a CI job for MySQL 8.4 and doesn't do anything for Docker etc. yet.

File: .github/workflows/unit_test_mysql84.yml [hunk 2]
```diff
@@ -0,0 +1,175 @@
+# DO NOT MODIFY: THIS FILE IS GENERATED USING "make generate_ci_workflows"
+
+name: Unit Test (mysql84)
+on: [push, pull_request]
+concurrency:
+  group: format('{0}-{1}', ${{ github.ref }}, 'Unit Test (mysql84)')
+  cancel-in-progress: true
```
Comment (frouioui) [inline .github/workflows/unit_test_mysql84.yml:7]: FYI, we need to mark this workflow as required once the PR is merged.

**Review by frouioui — APPROVED** [review]

**Review by vmg — APPROVED** [review]

File: go/mysql/flavor_mysqlgr.go [hunk 3]
```diff
@@ -42,8 +42,8 @@ type mysqlGRFlavor struct {
 }
 
 // newMysqlGRFlavor creates a new mysqlGR flavor.
-func newMysqlGRFlavor() flavor {
-	return &mysqlGRFlavor{}
+func newMysqlGRFlavor(serverVersion string) flavor {
```
Comment (deepthi) [inline go/mysql/flavor_mysqlgr.go:45]: We should probably delete this file? WDYT?

**Review by deepthi — COMMENTED** [review]
Nice work!

**Pull request merged**